### PR TITLE
feat(fxa-45): Add Fx Desktop states needed for device registration.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
This goes along with @philbooth's work in #69 

@rfk, @philbooth, @mhammond, and @ncalexan - This is by no means complete but a starting point for discussion. Mind giving a read and some feedback?

### CODE REVIEW

File: features/FxA-45-device-registration-api/FX-DESKTOP.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+Firefox Device Registration
+===========================
+
+This is a breakdown of the states Firefox needs to handle/worry about.
+
+See the [README.md](README.md) file for a list of requirements and user
```
Comment (shane-tomlinson) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: This file will exist once #69 is merged.

File: features/FxA-45-device-registration-api/FX-DESKTOP.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+* deviceName
+* deviceType
+* pushCallback
+* pushPublicKey
+
+Firefox must then store the returned deviceID.
+
+### Registration failure
+
+If registration fails because of a network error, Firefox should re-attempt
```
Comment (mhammond) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: At first blush, I expect failure to register will be "fatal" in terms of Sync starting - ie, much like an auth error.

File: features/FxA-45-device-registration-api/FX-DESKTOP.md (hunk 1 above)
Comment (mhammond) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: I see no problem with this, but I do expect the devil to be hiding in some of the details :)

File: features/FxA-45-device-registration-api/FX-DESKTOP.md (hunk 2 above)
Comment (rfk) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: >  like an auth error

From a user-visible-errors point of view, it may be more like a "can't connect to the tokenserver" error, i.e. it prevents sync from syncing, but there's nothing the user can do to intervene.

File: features/FxA-45-device-registration-api/FX-DESKTOP.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+## Existing profile, has FxA deviceID, has FxA sessionToken
+
+This is a user who has previously signed in to FxA to
+connect to Sync and their device is already registered.
+
+This device has already sent its initial name to the FxA
+auth server, but the user may have remotely disconnected
+the device, remotely reset the account password, or remotely
+changed the account password. The FxA auth server is considered
+the canonical source of truth for device state.
```
Comment (rfk) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: You'll eventually be able to edit the device-name in both places, so I'm not sure "server-canonical" is the right way to think about it.  Would this mean that e.g. you can't edit the device-name in chrome UI when it's offline and can't save it to the server?

File: features/FxA-45-device-registration-api/FX-DESKTOP.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+### Device list fetch failure
+
+If device list fails because of a network error, Firefox should re-attempt
+to re-fetch the list after a TBD interval has elapsed.
+
+## Firefox receives an `fxaccounts:login` message
+
+A user has just authenticated with Firefox Accounts. The message payload
+will contain both a sessionToken.
```
Comment (rfk) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: "both a sessionToken and...?" :-)

File: features/FxA-45-device-registration-api/FX-DESKTOP.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+If Firefox already has an FxA deviceID, then the user re-authenticated
+for some reason (reset password, change password).
+
+XXX Will the devices API automatically re-associate the old deviceID with the
+new sessionToken when the user signs in? Is there any mechanism for the auth
+server to do that automatically?
```
Comment (rfk) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: Yes, we should try.  If the device provides its deviceid when requesting a re-auth, then the auth-server will receive it in the call to /account/login and should *replace* any existing session-token with the newly-generated one.

File: features/FxA-45-device-registration-api/FX-DESKTOP.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+If Firefox does not have an FxA deviceID, then the device should be registered.
+See the section titled `Existing profile, no FxA deviceID, has FxA
+sessionToken`
+
+## Firefox receives an `fxaccounts:logout` message
+
+The user has just signed out of Firefox Accounts by removing the current device. The stored sessionToken and deviceID should be deleted, and the browser should disconnect from Sync.
```
Comment (rfk) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: In fact I think we should be a little more explicit here and state that the act of "disconnecting from sync", no matter how it's triggered, should involve destroying the device record and session token and discarding sessionToken and deviceID.

File: features/FxA-45-device-registration-api/FX-DESKTOP.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
+Firefox Device Registration
+===========================
+
+This is a breakdown of the states Firefox needs to handle/worry about.
```
Comment (rfk) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: It would also be useful to add a description of what the desired "steady state" is, i.e. that in the absence of any changes a connected device should:

* have a deviceID and a sessionToken
* use its deviceID to identify itself in the sync clients collection
* have a name that's consistent in local storage, fxa server record, and sync clients record

File: features/FxA-45-device-registration-api/FX-DESKTOP.md (hunk 5 above)
Comment (shane-tomlinson) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: > the auth-server will receive it in the call to /account/login

This is what I expected, and to make this possible, some method of passing `deviceID` to the content server will need to exist or else we cannot call `/account/login` w/ the existing `deviceID`. 

Another thought I had is to have the browser do all the work - it could call POST `/account/device` with the new `sessionToken`, the existing `name` and `deviceID`. The auth server could pull then associate that device with the new `sessionToken`?

File: features/FxA-45-device-registration-api/FX-DESKTOP.md (hunk 4 above)
Comment (shane-tomlinson) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: yyyeeesss..  doh.

File: features/FxA-45-device-registration-api/FX-DESKTOP.md (hunk 3 above)
Comment (shane-tomlinson) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: >  Would this mean that e.g. you can't edit the device-name in chrome UI when it's offline and can't save it to the server?

Realistically, the "offline" argument is a hard sell.

The situation I was thinking of when I wrote this involves remote device renaming. Say I have two devices w/ ids `device A` and `device B`. `device A` is `offline` or Fx is just shut down. On `device A`, I change the name to `Awesome device A`. I'm a wise-guy and want to see what happens, so on `device B`, which is online, I rename `device A` to `Awesome device A+` after I've done the rename on `device A` but before `device A` connects to the internet.

If we want to handle offline renaming, we'd have to keep track of an extra bit of information, "was the device renamed while offline?" If so, then ignore the server name and instead update the server. Or, store the date/time the name was changed, last change wins. Seems like more effort and possible brittleness than is worthwhile.

This does bring up an interesting question though, in a pre-push notification world, should a query for device name change be done on some regular interval? If we add push support immediately after v1, maybe not.

File: features/FxA-45-device-registration-api/FX-DESKTOP.md (hunk 5 above)
Comment (rfk) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: Yep.  I see three broad approaches:

1. Have the device pass it as a query parameter, e.g. `context=fx_desktop_v2&device=ABCDEF` and maybe even `device=` to signal that it has no id, but wants one.
2. Coordinate the deviceid as part of the device handshake.
3. Don't worry about it in the content-server at all, and just have the device bootstrap itself into getting a device-id once it has a session token.

The *eventual* right thing is probably (2), but I don't have strong feelings about whether (1) or (3) would work better for the initial version.

There are hypothetical reasons why the auth-server might like to know whether it's allocating a device-session or web-session token, e.g. https://github.com/mozilla/fxa-auth-server/issues/286, but YAGNI could easily overpower them.

File: features/FxA-45-device-registration-api/FX-DESKTOP.md (hunk 3 above)
Comment (rfk) [inline features/FxA-45-device-registration-api/FX-DESKTOP.md]: > Seems like more effort and possible brittleness than is worthwhile.

I broadly agree, we shouldn't get too caught up on this case, although we might be able to do something simple like have an etag on the device-description endpoint.  IIUC the browser currently uses etags to help manage the user's profile picture.

>  in a pre-push notification world, should a query for device name change be done on some regular interval?

Let's copy whatever we currently do for the profile picture, and move very quickly to add push support after v1.

**Comment by vladikoff** [comment]
@shane-tomlinson from mtg: should we just merge this with open questions and create another PR later? 

Or do you need more `r?`

**Comment by rfk** [comment]
FWIW, I think it'd be fine to merge it as a draft and iterate in smaller follow-up issues, these things can get a bit hard to follow once they're past a certain size.
